### DISCUSSION TITLE
How can a node discover another once they join a network?

### DISCUSSION THREAD

**Discussion by dankore** [Nephys/plexus#2] [General]
Let's say I have 3 nodes running, if a 4th node joins the network, how could the first 3 nodes automatically get the new node's IP address?

**Comment by Nephys** [comment]
Hi, for a node to connect to the network it has to first connect to another
node.

Let's say your 4th node connects to node 1, node 1 is connected to both
node 2 and 3, node 2 and 3 are not directly connected, but they are
indirectly connected through node 1.
Each node has an ID (in Hex, like 0xSomething);

So far you have 4 nodes online, all of  them are connected to node 1 but
not to each other.
But each node can ask any other node "hey whats the ip of node 4", you can
do so by using the find function of a node and set the key parameter to
node 4's ID (not IP).
Your node will look through the list of nodes they are directly connected
to, if they find a node with the specified ID it will contact it otherwise
it will pick a handful of nodes it thinks are close to the one you are
looking for and ask them if they know node 4.

It's like trying to find someone but you only know the name of the street
they live on, you pick someone who lives there too and ask them if they
know someone named node 4 (ID), if they do they will tell you node 4 lives
there (giving you their IP)
if they don't they will ask someone else who might live closer if they know
node 4, so on and so forth.

I hope this explanation is clear enough, I'm not that good at explaining
things, you might want to have a look at the file "/tests/test.js", line 76
"Node lookup".

Thank you for your time and thank you for trying my code.

**Reply by dankore** [reply]
Would you be interested in helping implement this library into a decentralized currency like Bitcoin? I'm building it using TypesScript.

It is not near as complicated as Bitcoin. I have gotten the basic currency working now except for the p2p network. 

If the currency picks up, so will your lib. I have built a small community around the currency and people are excited because I'm naming it after a town in Nigeria.

Right now, I'm using a central DB (Redis) to broadcast blocks and transactions instead of p2p.

In any case, I will be trying out your lib to do that. The repo to the currency I'm building is private but would be happy to add you to the project to take a look before you commit if you are interested.

**Reply by Nephys** [reply]
Hey, sorry for the late reply, been really busy lately.

Everything you need to implement this library should be available in the docs, but I'd be glad to help whenever I can.
I can't commit to such a project though because I'm currently working on other projects of my own as well as projects for a few clients I have.

Best of luck with your project and feel free to ask if you need any help with this library.
